src/core/agent/services/connectionService.ts

```diff
@@ -223,6 +219,17 @@ class ConnectionService extends AgentService {
         notes.push(JSON.parse(connection[key]));
       }
     });
+
+    const historyItems: Array<IpexMessageRecord> = [];
+    Object.keys(connection).forEach((key) => {
```

Comment:
Instead of iterating over the entire connection again, we should iterate over it once, and populate the connection notes and IPEX history items based on what matches in one go